gateway/gateway.go

```diff
@@ -84,43 +99,22 @@ func (g Gateway) HandleServiceRequest(
 		dataReporter:        g.DataReporter,
 	}
 
-	defer func() {
-		// Broadcast all observations, e.g. protocol-level, QoS-level, etc. contained in the gateway request context.
-		gatewayRequestCtx.BroadcastAllObservations()
-	}()
-
 	// Initialize the GatewayRequestContext struct using the HTTP request.
 	// e.g. extract the target service ID from the HTTP request.
 	err := gatewayRequestCtx.InitFromHTTPRequest(httpReq)
 	if err != nil {
 		return
 	}
 
-	// Determine the type of service request and handle it accordingly.
-	switch determineServiceRequestType(httpReq) {
-	case websocketServiceRequest:
-		g.handleWebSocketRequest(ctx, httpReq, gatewayRequestCtx, responseWriter)
-	default:
-		g.handleHTTPServiceRequest(ctx, httpReq, gatewayRequestCtx, responseWriter)
-	}
-}
-
-// handleHTTPServiceRequest handles a standard HTTP service request.
-func (g Gateway) handleHTTPServiceRequest(
-	_ context.Context,
-	httpReq *http.Request,
-	gatewayRequestCtx *requestContext,
-	responseWriter http.ResponseWriter,
-) {
-	logger := g.Logger.With("method", "handleHTTPServiceRequest")
-	defer func() {
+	defer func() { // Broadcast all observations, e.g. protocol-level, QoS-level, etc. contained in the gateway request context.
+		gatewayRequestCtx.BroadcastAllObservations()
```

Comment:
Writing user's response should be done first: both for faster response times and for potentially generating metrics related to response writing.